File: web/src/api/codegen.ts
import request from '@/utils/request'
import type { ApiResponse } from '@/types/api'

// ==================== 代码生成 类型定义 ====================

/** 表信息 */
export interface TableInfo {
  tableName: string
  tableComment: string
  engine: string
  createTime: string
  updateTime: string
}

/** 列信息 */
export interface ColumnInfo {
  columnName: string
  columnType: string
  columnComment: string
  isPk: boolean
  isNull: boolean
  maxLength: number
  columnDefault: string
  extra: string
}

/** 字段配置 */
export interface ColumnConfig {
  columnName: string
  columnType: string
  goType: string
  goField: string
  tsType: string
  label: string
  htmlType: string
  queryType: string
  isList: boolean
  isQuery: boolean
  isRequired: boolean
  isEdit: boolean
  dictType: string
  comment: string
  isPk: boolean
  isNull: boolean
  maxLength: number
  sort: number
}

/** 代码生成配置 */
export interface CodegenConfig {
  id?: number
  tableName: string
  tableComment: string
  className: string
  businessName: string
  functionName: string
  moduleName: string
  packageName: string
  author: string
  fields: ColumnConfig[]
  createdAt?: string
}

/** 代码预览请求 */
export interface CodegenPreviewRequest {
  tableName: string
  tableComment: string
  className: string
  businessName: string
  functionName: string
  moduleName: string
  packageName: string
  author: string
  fields: ColumnConfig[]
}

/** 代码文件 */
export interface CodegenFile {
  fileName: string
  filePath: string
  content: string
}

/** 代码预览响应 */
export interface CodegenPreviewResponse {
  files: CodegenFile[]
}

// ==================== 代码生成 API ====================

/** 获取数据库表列表 */
export function getTableList() {
  return request.get<ApiResponse<TableInfo[]>>('/codegen/tables')
}

/** 获取表的列信息 */
export function getColumnList(tableName: string) {
  return request.get<ApiResponse<ColumnInfo[]>>(`/codegen/columns/${tableName}`)
}

/** 代码预览 */
export function previewCode(data: CodegenPreviewRequest) {
  return request.post<ApiResponse<CodegenPreviewResponse>>('/codegen/preview', data)
}

/** 生成代码并下载 */
export function generateCode(data: CodegenPreviewRequest) {
  return request.post('/codegen/generate', data, { responseType: 'blob' })
}

/** 保存代码生成配置 */
export function saveConfig(data: CodegenConfig) {
  return request.post<ApiResponse<null>>('/codegen/config', data)
}

/** 获取代码生成配置 */
export function getConfig(tableName: string) {
  return request.get<ApiResponse<CodegenConfig>>(`/codegen/config/${tableName}`)
}

/** 获取所有代码生成配置 */
export function getAllConfigs() {
  return request.get<ApiResponse<CodegenConfig[]>>('/codegen/configs')
}

/** 删除代码生成配置 */
export function deleteConfig(id: number) {
  return request.delete<ApiResponse<null>>(`/codegen/config/${id}`)
}

File: server/model/response/codegen.go
package response

// TableInfoResponse 数据库表信息响应
type TableInfoResponse struct {
	TableName    string `json:"tableName"`    // 表名
	TableComment string `json:"tableComment"` // 表注释
	Engine       string `json:"engine"`       // 存储引擎
	CreateTime   string `json:"createTime"`   // 创建时间
}

// ColumnInfoResponse 数据库列信息响应
type ColumnInfoResponse struct {
	ColumnName    string `json:"columnName"`    // 列名
	ColumnType    string `json:"columnType"`    // 列类型
	ColumnComment string `json:"columnComment"` // 列注释
	IsPk          bool   `json:"isPk"`          // 是否主键
	IsNull        bool   `json:"isNull"`        // 是否可空
	MaxLength     int    `json:"maxLength"`     // 最大长度
	GoType        string `json:"goType"`        // Go 类型
	GoField       string `json:"goField"`       // Go 字段名
	TsType        string `json:"tsType"`        // TS 类型
	HtmlType      string `json:"htmlType"`      // 表单类型
	QueryType     string `json:"queryType"`     // 查询方式
	IsList        bool   `json:"isList"`        // 是否列表显示
	IsQuery       bool   `json:"isQuery"`       // 是否查询条件
	IsRequired    bool   `json:"isRequired"`    // 是否必填
	IsEdit        bool   `json:"isEdit"`        // 是否可编辑
	DictType      string `json:"dictType"`      // 字典类型
	Sort          int    `json:"sort"`          // 排序
}

// CodegenPreviewResponse 代码预览响应
type CodegenPreviewResponse struct {
	Files []CodegenFile `json:"files"` // 生成的文件列表
}

// CodegenFile 代码文件
type CodegenFile struct {
	FileName string `json:"fileName"` // 文件名
	FilePath string `json:"filePath"` // 文件路径
	Content  string `json:"content"`  // 文件内容
}

// CodegenConfigResponse 代码生成配置响应
type CodegenConfigResponse struct {
	ID           int64          `json:"id"`           // 配置ID
	TableName    string         `json:"tableName"`    // 表名
	TableComment string         `json:"tableComment"` // 表注释
	ClassName    string         `json:"className"`    // 类名
	BusinessName string         `json:"businessName"` // 业务名
	FunctionName string         `json:"functionName"` // 功能名
	ModuleName   string         `json:"moduleName"`   // 模块名
	PackageName  string         `json:"packageName"`  // 包名
	Author       string         `json:"author"`       // 作者
	Fields       []ColumnConfig `json:"fields"`       // 字段配置
	CreatedAt    string         `json:"createdAt"`    // 创建时间
}

// ColumnConfig 字段配置（用于配置响应）
type ColumnConfig struct {
	ColumnName string `json:"columnName"` // 数据库列名
	ColumnType string `json:"columnType"` // 数据库类型
	GoType     string `json:"goType"`     // Go 类型
	GoField    string `json:"goField"`    // Go 字段名
	TsType     string `json:"tsType"`     // TS 类型
	Label      string `json:"label"`      // 显示标签
	HtmlType   string `json:"htmlType"`   // 表单类型
	QueryType  string `json:"queryType"`  // 查询方式
	IsList     bool   `json:"isList"`     // 是否列表显示
	IsQuery    bool   `json:"isQuery"`    // 是否查询条件
	IsRequired bool   `json:"isRequired"` // 是否必填
	IsEdit     bool   `json:"isEdit"`     // 是否可编辑
	DictType   string `json:"dictType"`   // 字典类型
	Comment    string `json:"comment"`    // 字段注释
	IsPk       bool   `json:"isPk"`       // 是否主键
	IsNull     bool   `json:"isNull"`     // 是否可空
	MaxLength  int    `json:"maxLength"`  // 最大长度（用于 VARCHAR）
	Sort       int    `json:"sort"`       // 排序
}
